sql/job_stat_history_log_retention.sql

```diff
@@ -4,23 +4,110 @@
 
 -- A retention policy is set up for the table _timescaledb_internal.job_errors (Error Log Retention Policy [2])
 -- By default, it will run once a month and and drop rows older than a month.
+
+-- We use binary search on the id column to figure out which rows should be retained from the job history table
+-- Doing it this way allows us to use the index on the `id` column, which (empirically) we found is faster than querying on the `execution_finish` column directly without an index
+-- This works because `execution_finish` is always ordered.
+-- We can consider alternative approaches to simplify this in the future.
+CREATE OR REPLACE FUNCTION _timescaledb_functions.job_history_bsearch(search_point TIMESTAMPTZ) RETURNS BIGINT
+AS
+$$
+DECLARE
+  id_lower BIGINT;
+  id_upper BIGINT;
+  id_middle BIGINT DEFAULT 0;
+  target_tz TIMESTAMPTZ;
+BEGIN
+
+  SELECT COALESCE(min(id), 0), COALESCE(max(id), 0)
+  INTO id_lower, id_upper
+  FROM _timescaledb_internal.bgw_job_stat_history;
+
+  IF id_lower = 0 AND id_upper = 0 THEN
+    RETURN NULL;
+  END IF;
+
+  -- We want the first entry in the table where execution_finish is >= search_point
+  WHILE id_lower < id_upper LOOP
+    id_middle := id_lower + (id_upper - id_lower) / 2;
+
+    SELECT execution_finish
+    INTO target_tz
+    FROM _timescaledb_internal.bgw_job_stat_history
+    WHERE id = id_middle;
+
+    -- If the id_middle is not found, shift to a previous id that's still in the search space
+    IF NOT FOUND THEN
+      SELECT execution_finish, id
+      INTO target_tz, id_middle
+      FROM _timescaledb_internal.bgw_job_stat_history
+      WHERE id <= id_middle AND id >= id_lower
+      ORDER BY id LIMIT 1;
+
+      IF NOT FOUND THEN
+        id_middle := id_lower;
+      END IF;
+
+    END IF;
+
+    IF target_tz >= search_point THEN
+      id_upper := id_middle;
+    ELSE
+      id_lower := id_middle + 1;
+    END IF;
+  END LOOP;
+
+  -- Handle the case where no ids need to be deleted and return NULL instead
+  SELECT execution_finish
+  INTO target_tz
+  FROM _timescaledb_internal.bgw_job_stat_history
+  WHERE id = id_lower;
+
+  IF target_tz < search_point THEN
+    RETURN NULL;
+  END IF;
+
+  RETURN id_lower;
+END
+$$
+LANGUAGE plpgsql SET search_path TO pg_catalog, pg_temp;
+
 CREATE OR REPLACE FUNCTION _timescaledb_functions.policy_job_stat_history_retention(job_id integer, config JSONB) RETURNS integer
 LANGUAGE PLPGSQL AS
 $BODY$
 DECLARE
-    drop_after INTERVAL;
     numrows INTEGER;
+    search_point TIMESTAMPTZ;
+    id_found BIGINT;
 BEGIN
-    drop_after := config->>'drop_after';
+  SET LOCAL lock_timeout = '5s';
```

Comment:
```suggestion
  PERFORM set_config('lock_timeout', coalesce(config->>'lock_timeout', '5s'), true /* is local */);
```
We can add a test for it in a following PR. The idea is to be able to set a config parameter with `lock_timeout` key. Just wondering if for some reason we can have a problem in our fleet with the fixed 5 seconds default value.

Reply:
Updated the PR. Need to add a proper test for it. We should probably also add an isolation test to check the lock timeout works correctly. Can be a follow-up PR.